Kubernetes/legos/k8s_check_service_pvc_utilization/k8s_check_service_pvc_utilization.py

```diff
@@ -97,73 +97,103 @@ def k8s_check_service_pvc_utilization(handle, service_name: str = "", namespace:
-            mount_path = mount['mount_path']
-            pvc_name = mount['pvc_name']
-            all_pvcs.append({"pvc_name": pvc_name, "mount_path": mount_path, "used": None, "capacity": None})
-
-        all_mounts = [mount.get('mount_path') for mount in pvc_mounts]
-        all_mounts = " ".join(all_mounts).strip()
-        du_command = f"kubectl exec -n {namespace} {pod_name} -c {container_name} -- df -kh {all_mounts} | grep -v Filesystem"
-        du_output = handle.run_native_cmd(du_command)
-
-
-        if du_output and not du_output.stderr:
-            used_space = du_output.stdout.strip()
-            for idx, space in enumerate([used_space]):
-                space = space.split()
-                used_percentage = int(space[-2].replace('%', ''))
-                total_capacity_str = space[1].replace('%', '')
-                all_pvcs[idx]["used"] = used_percentage
-                all_pvcs[idx]["capacity"] =  total_capacity_str
-                if used_percentage > threshold:
-                    alert_pvcs.append(all_pvcs[idx])
-
-        alert_pvcs_all_services.extend(alert_pvcs)
+        for element in pod_and_pvc_names:
+            pod_name, claim_name = element.split(':')
+
+            # Fetch the Pod JSON 
+            # We need to get the container name (if any) from the Pod's JSON. This is needed
+            # if we want to exec into the POD that is within a container. The JSON data that
+            # we obtain is used to fill the pvc_mounts list, which is a list of dictionaries.
+            # We use this pvc_mounts to find out the used_space percentage. We compare that with
+            # the threshold to flag if the utilization is above threshold. 
+            # df -kh is the command used to get the disk utilization. This is accurate as we get
+            # the disk utilization from the POD directly, rather than checking the resource limit
+            # and resource request from the deployment / stateful YAML file. 
+            get_pod_json_command = f"kubectl get pod {pod_name} -n {namespace} -o json"
+            pod_json_output = handle.run_native_cmd(get_pod_json_command)
+            if not pod_json_output or pod_json_output.stderr:
+                raise ApiException(f"Error fetching pod json for {pod_name}: {pod_json_output.stderr if pod_json_output else 'empty response'}")
+            pod_data = json.loads(pod_json_output.stdout)
+    
+            pvc_mounts.extend([
+                {"container_name": container['name'],
+                "mount_path": mount['mountPath'],
+                "pvc_name": claim_name if claim_name else volume['persistentVolumeClaim']['claimName']}
+                for container in pod_data['spec']['containers']
+                for mount in container.get('volumeMounts', [])
+                for volume in pod_data['spec']['volumes']
+                if 'persistentVolumeClaim' in volume and volume['name'] == mount['name']
+            ])
```

Comment:
No `None` checks for any of the fields being accessed.. Are they guaranteed to be populated?

Reply:
Rewrote this with simple for loops with error handling for each of the loop